### ISSUE TITLE
Hangs with certain syntaxes

### ISSUE THREAD

**Issue by jgm** [jgm/skylighting#3]
Let allcases be the concatenation of the contents of all files in test/cases/.

Then, as demonstrated by our new test suite, skylighting hangs for Alert, Alert_indent,
D, Hamlet, Modelines, Maxima, Scheme.

```
% echo "  " | skylighting -s modelines --trace 
DetectSpaces MATCHED Just (CommentTok,"  ")
(hangs)

% echo "hi" | skylighting -s Alerts 
(hangs)

% echo "hi" | skylighting -s hamlet
(hangs)

% echo "hi" | skylighting -s maxima
(hangs)
```

With D, we get teh problem with this snippet:
```
foreach ($words as $word) {
    echo $word.': ';
    echo canMakeWord($word) ? "True" : "False";
```
Tracing shows this loop:
```
CONTEXT STACK ["normal"]
DetectSpaces MATCHED Just (NormalTok," ")
DetectChar MATCHED Just (CharTok,"'")
CONTEXT STACK ["CharLiteral","normal"]
RegExpr MATCHED Just (CharTok,";")
CONTEXT STACK ["CharLiteralClosing","CharLiteral","normal"]
CONTEXT STACK ["CharLiteral","normal"]
CONTEXT STACK ["CharLiteralClosing","CharLiteral","normal"]
CONTEXT STACK ["CharLiteral","normal"]
CONTEXT STACK ["CharLiteralClosing","CharLiteral","normal"]
CONTEXT STACK ["CharLiteral","normal"]
```

With scheme, we hang on
```
  -- AX<=< % AXsymm+    so r < r + 1
#include <stdio.h>
```
with trace:
```
CONTEXT STACK ["SpecialNumber","Level2","Level1","Level0"]
IncludeRules MATCHED Just (CharTok,"#i")
(hangs)
```

**Comment by jgm** [comment]
OK, most of these now fixed.

Outstanding issues:

maxima fails with Empty stack error on this maxima program:
```
  /* incomplete preliminary version  */
allOpsPriv(expression, opList) :=
 block ( [x],
        if atom(expression)
           then opList
           else
             (x: op(expression),
              cons(x, opList)
             ) 
        );
```
The rule for the Normal context has pop as line-end context shift.

Haxe is dropping some text.

And scheme has a timeout (same repro as above).

**Comment by jgm** [comment]
Fixed with 70dfbf28e8af72861b9f897b095443df9fd20b64 and earlier commits.
